src/wasm/wasm-emscripten.cpp

```diff
@@ -286,20 +291,40 @@ Function* EmscriptenGlueGenerator::generateAssignGOTEntriesFunction() {
     if (ex) {
       assert(ex->kind == ExternalKind::Function);
       f = wasm.getFunction(ex->value);
-    } else {
-      ImportInfo info(wasm);
-      f = info.getImportedFunction(ENV, g->base);
-      if (!f) {
-        Fatal() << "GOT.func entry with no import/export: " << g->base;
+      if (!isSideModule) {
+        // This is exported, so must be one of the functions implemented here.
+        // Simply add it to the table, and use that index. The loader will
+        // know to reuse that index for other modules so they all share the
+        // same index and function pointer equality works.
+        if (f->imported()) {
+          Fatal() << "GOT.func entry export but not implemented: " << g->base;
+        }
+        auto tableIndex = TableUtils::getOrAppend(wasm.table, f->name, wasm);
+        auto* c = LiteralUtils::makeFromInt32(tableIndex, Type::i32, wasm);
+        // The base relative to which we are computed is the offset of the
+        // singleton segment.
+        auto* getBase =
+          ExpressionManipulator::copy(wasm.table.segments[0].offset, wasm);
+        auto* add = builder.makeBinary(AddInt32, getBase, c);
+        auto* globalSet = builder.makeGlobalSet(g->name, add);
+        block->list.push_back(globalSet);
```

Comment:
I think we may have forgotten to take into account `__table_base` here, and also the `tablesize` in the `dylink` section?  

If we create the table during this pass we will most likely also need to create `__table_base` which would otherwise have been created by lld.

Reply:
See the `::copy` operation: we are copying the base of the segment, and adding that. The tests show we handle the case of `__table_base` as well as if it's a constant.

If we create the table here, then we create it with a constant offset. I think that won't happen in dynamic linking since we will always import the table there, though?

About `dylink`, yeah, thinking on that now, it seems like we need to update it if it exists. Looks like we have no handling of that at all in binaryen yet.

What do you think about doing this in binaryen vs wasm-ld? After your comments here it is starting to seem to me like wasm-ld might be simpler (but I don't know the code there).

Reply:
I thought about it and I think doing in lld would require more re-design.  I'm OK with landing this change first/now, while I look at perhaps trying do this in wasm-ld.

Reply:
Sounds good to do it here.

Meanwhile I looked into the dylink section issue. It would require first adding full parsing and writing support for that section, and for updating it through operations, so basically adding IR support for it. If we want that we should do it separately from this PR I think. And that doesn't need to block this PR as this one only works on main modules where there is no dylink section.

Reply:
The MAIN_MODULE does have a dylink section.  We should probably fix binaryen to preserve that section by I agree that is a separate PR.

Reply:
Oh interesting, that's a difference in upstream compared to fastcomp that I didn't realize.

Emscripten never uses it in a main module, though, I verified.